File: src/scope.js
const fs = require("node:fs");
const path = require("node:path");

const CONFIG_FILENAME = ".flowpeek/config.json";
const CONFIG_SCHEMA_VERSION = 1;
const DEFAULT_SCOPE_CONFIG = Object.freeze({
  schemaVersion: CONFIG_SCHEMA_VERSION,
  sourceRoots: [],
  testRoots: ["test", "tests", "__tests__"],
  fixtureRoots: ["test/fixtures", "tests/fixtures", "__fixtures__"],
  exclude: [],
  projectId: null,
  flowEntries: { tests: false, fixtures: false },
});
const GENERATED_DIRECTORY_NAMES = new Set(["generated", "__generated__"]);
const PROJECT_ID_ALLOWED_CHARACTERS = new Set("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789:._-");

class RepositoryScopeConfigError extends Error {
  constructor(message) {
    super(`Invalid Flowpeek repository scope configuration: ${message}`);
    this.name = "RepositoryScopeConfigError";
    this.code = "FLOWPEEK_INVALID_SCOPE_CONFIG";
  }
}

function toPosix(value) {
  return String(value).replaceAll("\\", "/");
}

function splitPath(value) {
  return toPosix(value).split("/").filter(Boolean);
}

function normaliseRule(value, field, { allowGlob = false } = {}) {
  if (typeof value !== "string" || !value.trim()) throw new RepositoryScopeConfigError(`${field} entries must be non-empty strings.`);
  const normalized = toPosix(value.trim()).replace(/^\.\/+/, "").replace(/\/+$/, "") || ".";
  if (path.isAbsolute(value) || path.win32.isAbsolute(value) || normalized.startsWith("/")) throw new RepositoryScopeConfigError(`${field} entries must be repository-relative paths.`);
  const segments = splitPath(normalized);
  if (segments.includes("..")) throw new RepositoryScopeConfigError(`${field} entries must not traverse outside the repository.`);
  if (!allowGlob && segments.some((segment) => segment.includes("*"))) throw new RepositoryScopeConfigError(`${field} roots cannot contain glob segments.`);
  if (allowGlob && segments.some((segment) => segment.includes("*") && segment !== "*" && segment !== "**")) throw new RepositoryScopeConfigError(`${field} supports only whole-segment * and ** globs.`);
  return normalized;
}

function normalisePathList(value, field, options) {
  if (!Array.isArray(value)) throw new RepositoryScopeConfigError(`${field} must be an array of repository-relative paths.`);
  return [...new Set(value.map((entry) => normaliseRule(entry, field, options)))].sort();
}

function cloneDefaults() {
  return {
    schemaVersion: CONFIG_SCHEMA_VERSION,
    sourceRoots: [...DEFAULT_SCOPE_CONFIG.sourceRoots],
    testRoots: [...DEFAULT_SCOPE_CONFIG.testRoots],
    fixtureRoots: [...DEFAULT_SCOPE_CONFIG.fixtureRoots],
    exclude: [...DEFAULT_SCOPE_CONFIG.exclude],
    projectId: DEFAULT_SCOPE_CONFIG.projectId,
    flowEntries: { ...DEFAULT_SCOPE_CONFIG.flowEntries },
  };
}

function isValidProjectId(value) {
  if (typeof value !== "string" || value.length < 8 || value.length > 160) return false;
  for (const character of value) {
    if (!PROJECT_ID_ALLOWED_CHARACTERS.has(character)) return false;
  }
  return true;
}

function validateScopeConfig(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new RepositoryScopeConfigError("config.json must contain an object.");
  const allowed = new Set(["schemaVersion", "sourceRoots", "testRoots", "fixtureRoots", "exclude", "projectId", "flowEntries"]);
  for (const key of Object.keys(value)) {
    if (!allowed.has(key)) throw new RepositoryScopeConfigError(`unknown property \"${key}\".`);
  }
  if (value.schemaVersion !== CONFIG_SCHEMA_VERSION) throw new RepositoryScopeConfigError(`schemaVersion must be ${CONFIG_SCHEMA_VERSION}.`);
  const defaults = cloneDefaults();
  const flowEntries = value.flowEntries === undefined ? defaults.flowEntries : value.flowEntries;
  if (!flowEntries || typeof flowEntries !== "object" || Array.isArray(flowEntries)) throw new RepositoryScopeConfigError("flowEntries must be an object.");
  for (const key of Object.keys(flowEntries)) {
    if (key !== "tests" && key !== "fixtures") throw new RepositoryScopeConfigError(`flowEntries has unknown property \"${key}\".`);
    if (typeof flowEntries[key] !== "boolean") throw new RepositoryScopeConfigError(`flowEntries.${key} must be a boolean.`);
  }
  const projectId = value.projectId === undefined ? defaults.projectId : value.projectId;
  if (projectId !== null && !isValidProjectId(projectId)) throw new RepositoryScopeConfigError("projectId must be null or a stable identifier containing only letters, numbers, colon, dot, underscore, or hyphen.");
  return {
    schemaVersion: CONFIG_SCHEMA_VERSION,
    sourceRoots: value.sourceRoots === undefined ? defaults.sourceRoots : normalisePathList(value.sourceRoots, "sourceRoots"),
    testRoots: value.testRoots === undefined ? defaults.testRoots : normalisePathList(value.testRoots, "testRoots"),
    fixtureRoots: value.fixtureRoots === undefined ? defaults.fixtureRoots : normalisePathList(value.fixtureRoots, "fixtureRoots"),
    exclude: value.exclude === undefined ? defaults.exclude : normalisePathList(value.exclude, "exclude", { allowGlob: true }),
    projectId,
    flowEntries: {
      tests: flowEntries.tests === undefined ? defaults.flowEntries.tests : flowEntries.tests,
      fixtures: flowEntries.fixtures === undefined ? defaults.flowEntries.fixtures : flowEntries.fixtures,
    },
  };
}

function readRepositoryScope(root) {
  const configPath = path.join(root, CONFIG_FILENAME);
  if (!fs.existsSync(configPath)) return { ...cloneDefaults(), source: "defaults", configPath: null };
  let parsed;
  try {
    parsed = JSON.parse(fs.readFileSync(configPath, "utf8"));
  } catch (error) {
    throw new RepositoryScopeConfigError(`.flowpeek/config.json is not valid JSON (${error.message}).`);
  }
  return { ...validateScopeConfig(parsed), source: "config", configPath: CONFIG_FILENAME };
}

function pathWithinRoot(relativePath, root) {
  if (root === ".") return true;
  return relativePath === root || relativePath.startsWith(`${root}/`);
}

function segmentMatches(pattern, value) {
  return pattern === "*" || pattern === value;
}

function matchesExcludePattern(relativePath, pattern) {
  const pathSegments = splitPath(relativePath);
  const patternSegments = splitPath(pattern);
  const visit = (pathIndex, patternIndex) => {
    if (patternIndex === patternSegments.length) return pathIndex === pathSegments.length;
    const token = patternSegments[patternIndex];
    if (token === "**") {
      if (patternIndex === patternSegments.length - 1) return true;
      for (let index = pathIndex; index <= pathSegments.length; index += 1) {
        if (visit(index, patternIndex + 1)) return true;
      }
      return false;
    }
    return pathIndex < pathSegments.length && segmentMatches(token, pathSegments[pathIndex]) && visit(pathIndex + 1, patternIndex + 1);
  };
  if (!patternSegments.some((segment) => segment.includes("*"))) return pathWithinRoot(relativePath, pattern);
  return visit(0, 0);
}

function looksLikeTestPath(relativePath) {
  const segments = splitPath(relativePath).map((segment) => segment.toLowerCase());
  const filename = segments.at(-1) || "";
  const extension = filename.lastIndexOf(".");
  const stem = extension > 0 ? filename.slice(0, extension) : filename;
  return segments.includes("__tests__") || stem.includes(".test") || stem.includes(".spec") || stem.endsWith("_test");
}

function looksGenerated(relativePath) {
  const segments = splitPath(relativePath).map((segment) => segment.toLowerCase());
  const filename = segments.at(-1) || "";
  return segments.some((segment) => GENERATED_DIRECTORY_NAMES.has(segment)) || filename.includes(".generated.");
}

function classifyRepositoryPath(relativePath, scope) {
  const normalized = normaliseRule(relativePath, "repository path");
  if (scope.exclude.some((pattern) => matchesExcludePattern(normalized, pattern))) return "excluded";
  if (scope.fixtureRoots.some((root) => pathWithinRoot(normalized, root))) return "fixture";
  if (scope.testRoots.some((root) => pathWithinRoot(normalized, root)) || looksLikeTestPath(normalized)) return "test";
  if (looksGenerated(normalized)) return "generated";
  if (scope.sourceRoots.length && !scope.sourceRoots.some((root) => pathWithinRoot(normalized, root))) return "excluded";
  return "application";
}

function scopeSummary(scope, records, excludedPaths = []) {
  const counts = { application: 0, test: 0, fixture: 0, generated: 0, excluded: excludedPaths.length };
  for (const record of records) counts[record.sourceScope] += 1;
  return {
    schemaVersion: CONFIG_SCHEMA_VERSION,
    source: scope.source,
    configPath: scope.configPath,
    sourceRoots: scope.sourceRoots,
    testRoots: scope.testRoots,
    fixtureRoots: scope.fixtureRoots,
    exclude: scope.exclude,
    projectId: scope.projectId,
    flowEntries: scope.flowEntries,
    precedence: ["excluded", "fixture", "test", "generated", "application"],
    counts,
    limitations: [
      "Scope classification is path-based and does not execute repository configuration.",
      "Generated source is retained for diagnostics but is not an application-flow entry by default.",
      "Diagnostic/all views include test, fixture, and generated source; their relationships remain static evidence, not runtime behavior.",
    ],
  };
}

function scopeSignature(scope) {
  return JSON.stringify({
    schemaVersion: scope.schemaVersion,
    sourceRoots: scope.sourceRoots,
    testRoots: scope.testRoots,
    fixtureRoots: scope.fixtureRoots,
    exclude: scope.exclude,
    projectId: scope.projectId,
    flowEntries: scope.flowEntries,
  });
}

module.exports = {
  CONFIG_FILENAME,
  CONFIG_SCHEMA_VERSION,
  DEFAULT_SCOPE_CONFIG,
  RepositoryScopeConfigError,
  classifyRepositoryPath,
  readRepositoryScope,
  scopeSignature,
  scopeSummary,
};

File: native/flopeek-core/src/scope.rs
use serde_json::Value;
use std::collections::BTreeSet;
use std::fs;
use std::path::Path;

pub const SCOPE_CONFIG_RELATIVE_PATH: &str = ".flopeek/config.json";
const CONFIG_SCHEMA_VERSION: i64 = 1;
const DEFAULT_TEST_ROOTS: &[&str] = &["test", "tests", "__tests__"];
const DEFAULT_FIXTURE_ROOTS: &[&str] = &["test/fixtures", "tests/fixtures", "__fixtures__"];

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SourceScope {
    Application,
    Test,
    Fixture,
    Generated,
    Excluded,
}

impl SourceScope {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Application => "application",
            Self::Test => "test",
            Self::Fixture => "fixture",
            Self::Generated => "generated",
            Self::Excluded => "excluded",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NativeScope {
    pub source: String,
    pub project_id: Option<String>,
    source_roots: Vec<String>,
    test_roots: Vec<String>,
    fixture_roots: Vec<String>,
    exclude: Vec<String>,
}

fn normalise_rule(value: &str, field: &str, allow_glob: bool) -> Result<String, String> {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return Err(format!("{field} entries must be non-empty strings."));
    }
    if trimmed.starts_with('/')
        || trimmed.starts_with("\\\\")
        || (trimmed.len() > 1 && trimmed.as_bytes()[1] == b':')
    {
        return Err(format!(
            "{field} entries must be repository-relative paths."
        ));
    }
    let mut normalized = trimmed.replace('\\', "/");
    while normalized.starts_with("./") {
        normalized = normalized[2..].to_string();
    }
    normalized = normalized.trim_end_matches('/').to_string();
    if normalized.is_empty() {
        normalized = ".".to_string();
    }
    let segments = normalized
        .split('/')
        .filter(|segment| !segment.is_empty())
        .collect::<Vec<_>>();
    if segments.iter().any(|segment| *segment == "..") {
        return Err(format!(
            "{field} entries must not traverse outside the repository."
        ));
    }
    for segment in &segments {
        if !segment.contains('*') {
            continue;
        }
        if !allow_glob || (*segment != "*" && *segment != "**") {
            return Err(format!(
                "{field} supports only whole-segment * and ** globs."
            ));
        }
    }
    Ok(normalized)
}

fn path_list(
    value: Option<&Value>,
    field: &str,
    defaults: &[&str],
    allow_glob: bool,
) -> Result<Vec<String>, String> {
    let values = match value {
        None => defaults
            .iter()
            .map(|value| Value::String((*value).to_string()))
            .collect::<Vec<_>>(),
        Some(Value::Array(values)) => values.clone(),
        Some(_) => {
            return Err(format!(
                "{field} must be an array of repository-relative paths."
            ));
        }
    };
    let mut normalized = BTreeSet::new();
    for value in values {
        let Some(value) = value.as_str() else {
            return Err(format!("{field} entries must be non-empty strings."));
        };
        normalized.insert(normalise_rule(value, field, allow_glob)?);
    }
    Ok(normalized.into_iter().collect())
}

fn valid_project_id(value: &str) -> bool {
    value.len() >= 8
        && value.len() <= 160
        && value.chars().all(|character| {
            character.is_ascii_alphanumeric() || matches!(character, ':' | '.' | '_' | '-')
        })
}

fn path_within_root(path: &str, root: &str) -> bool {
    root == "." || path == root || path.starts_with(&format!("{root}/"))
}

fn matches_pattern(path_segments: &[&str], pattern_segments: &[&str]) -> bool {
    match pattern_segments.split_first() {
        None => path_segments.is_empty(),
        Some((&"**", rest)) => {
            (0..=path_segments.len()).any(|index| matches_pattern(&path_segments[index..], rest))
        }
        Some((&pattern, rest)) => path_segments
            .split_first()
            .is_some_and(|(path, remaining)| {
                (pattern == "*" || pattern == *path) && matches_pattern(remaining, rest)
            }),
    }
}

fn matches_exclude(path: &str, pattern: &str) -> bool {
    if !pattern.contains('*') {
        return path_within_root(path, pattern);
    }
    matches_pattern(
        &path
            .split('/')
            .filter(|segment| !segment.is_empty())
            .collect::<Vec<_>>(),
        &pattern
            .split('/')
            .filter(|segment| !segment.is_empty())
            .collect::<Vec<_>>(),
    )
}

fn looks_like_test(path: &str) -> bool {
    let segments = path
        .to_ascii_lowercase()
        .split('/')
        .map(str::to_string)
        .collect::<Vec<_>>();
    let filename = segments.last().map(String::as_str).unwrap_or_default();
    let stem = filename
        .rsplit_once('.')
        .map(|(stem, _)| stem)
        .unwrap_or(filename);
    segments.iter().any(|segment| segment == "__tests__")
        || stem.contains(".test")
        || stem.contains(".spec")
        || stem.ends_with("_test")
}

fn looks_generated(path: &str) -> bool {
    let segments = path
        .to_ascii_lowercase()
        .split('/')
        .map(str::to_string)
        .collect::<Vec<_>>();
    let filename = segments.last().map(String::as_str).unwrap_or_default();
    segments
        .iter()
        .any(|segment| segment == "generated" || segment == "__generated__")
        || filename.contains(".generated.")
}

impl NativeScope {
    pub fn classify(&self, path: &str) -> SourceScope {
        if self
            .exclude
            .iter()
            .any(|pattern| matches_exclude(path, pattern))
        {
            return SourceScope::Excluded;
        }
        if self
            .fixture_roots
            .iter()
            .any(|root| path_within_root(path, root))
        {
            return SourceScope::Fixture;
        }
        if self
            .test_roots
            .iter()
            .any(|root| path_within_root(path, root))
            || looks_like_test(path)
        {
            return SourceScope::Test;
        }
        if looks_generated(path) {
            return SourceScope::Generated;
        }
        if !self.source_roots.is_empty()
            && !self
                .source_roots
                .iter()
                .any(|root| path_within_root(path, root))
        {
            return SourceScope::Excluded;
        }
        SourceScope::Application
    }
}

pub fn read_native_scope(root: &Path) -> Result<NativeScope, String> {
    let config_path = root.join(SCOPE_CONFIG_RELATIVE_PATH);
    if !config_path.exists() {
        return Ok(NativeScope {
            source: "defaults".to_string(),
            project_id: None,
            source_roots: Vec::new(),
            test_roots: DEFAULT_TEST_ROOTS
                .iter()
                .map(|value| (*value).to_string())
                .collect(),
            fixture_roots: DEFAULT_FIXTURE_ROOTS
                .iter()
                .map(|value| (*value).to_string())
                .collect(),
            exclude: Vec::new(),
        });
    }
    let content = fs::read_to_string(&config_path)
        .map_err(|error| format!("Unable to read .flopeek/config.json: {error}"))?;
    let value: Value = serde_json::from_str(&content)
        .map_err(|error| format!(".flopeek/config.json is not valid JSON ({error})."))?;
    let Some(object) = value.as_object() else {
        return Err(".flopeek/config.json must contain an object.".to_string());
    };
    for key in object.keys() {
        if !matches!(
            key.as_str(),
            "schemaVersion"
                | "sourceRoots"
                | "testRoots"
                | "fixtureRoots"
                | "exclude"
                | "projectId"
                | "flowEntries"
        ) {
            return Err(format!(
                ".flopeek/config.json has unknown property \"{key}\"."
            ));
        }
    }
    if object.get("schemaVersion").and_then(Value::as_i64) != Some(CONFIG_SCHEMA_VERSION) {
        return Err(".flopeek/config.json schemaVersion must be 1.".to_string());
    }
    let project_id = match object.get("projectId") {
        None | Some(Value::Null) => None,
        Some(Value::String(value)) if valid_project_id(value) => Some(value.clone()),
        Some(_) => {
            return Err(
                ".flopeek/config.json projectId must be null or a stable safe identifier."
                    .to_string(),
            );
        }
    };
    if let Some(flow_entries) = object.get("flowEntries") {
        let Some(flow_entries) = flow_entries.as_object() else {
            return Err(".flopeek/config.json flowEntries must be an object.".to_string());
        };
        for (key, value) in flow_entries {
            if !matches!(key.as_str(), "tests" | "fixtures") || !value.is_boolean() {
                return Err(".flopeek/config.json flowEntries is invalid.".to_string());
            }
        }
    }
    Ok(NativeScope {
        source: "config".to_string(),
        project_id,
        source_roots: path_list(object.get("sourceRoots"), "sourceRoots", &[], false)?,
        test_roots: path_list(
            object.get("testRoots"),
            "testRoots",
            DEFAULT_TEST_ROOTS,
            false,
        )?,
        fixture_roots: path_list(
            object.get("fixtureRoots"),
            "fixtureRoots",
            DEFAULT_FIXTURE_ROOTS,
            false,
        )?,
        exclude: path_list(object.get("exclude"), "exclude", &[], true)?,
    })
}
